Fix anon const def-creation when macros are involved

camelid opened the pull request:
Fixes #128016.

Ever since #125915, some `ast::AnonConst`s turn into `hir::ConstArgKind::Path`s,
which don't have associated `DefId`s. To deal with the fact that we don't have
resolution information in `DefCollector`, we decided to implement a process
where if the anon const *appeared* to be trivial (i.e., `N` or `{ N }`), we
would avoid creating a def for it in `DefCollector`. If later, in AST lowering,
we realized it turned out to be a unit struct literal, or we were lowering it
to something that didn't use `hir::ConstArg`, we'd create its def there.

However, let's say we have a macro `m!()` that expands to a reference to a free
constant `FOO`. If we use `m!()` in the body of an anon const (e.g., `Foo<{ m!() }>`),
then in def collection, it appears to be a nontrivial anon const and we create
a def. But the macro expands to something that looks like a trivial const arg,
but is not, so in AST lowering we "fix" the mistake we assumed def collection
made and create a def for it. This causes a duplicate definition ICE.

The long-term fix for this is to delay the creation of defs for all expression-like
nodes until AST lowering (see #128844 for an incomplete attempt at this). This
would avoid issues like this one that are caused by hacky workarounds. However,
doing this uncovers a pre-existing bug with opaque types that is quite involved
to fix (see #129023).

In the meantime, this PR fixes the bug by delaying def creation for anon consts
whose bodies are macro invocations until after we expand the macro and know
what is inside it. This is accomplished by adding information to create the
anon const's def to the data in `Resolver.invocation_parents`.

r? @BoxyUwU

camelid committed:
Fix anon const def-creation when macros are involved

Ever since #125915, some `ast::AnonConst`s turn into `hir::ConstArgKind::Path`s,
which don't have associated `DefId`s. To deal with the fact that we don't have
resolution information in `DefCollector`, we decided to implement a process
where if the anon const *appeared* to be trivial (i.e., `N` or `{ N }`), we
would avoid creating a def for it in `DefCollector`. If later, in AST lowering,
we realized it turned out to be a unit struct literal, or we were lowering it
to something that didn't use `hir::ConstArg`, we'd create its def there.

However, let's say we have a macro `m!()` that expands to a reference to a free
constant `FOO`. If we use `m!()` in the body of an anon const (e.g., `Foo<{ m!() }>`),
then in def collection, it appears to be a nontrivial anon const and we create
a def. But the macro expands to something that looks like a trivial const arg,
but is not, so in AST lowering we "fix" the mistake we assumed def collection
made and create a def for it. This causes a duplicate definition ICE.

The ideal long-term fix for this is a bit unclear. One option is to delay def
creation for all expression-like nodes until AST lowering (see #128844 for an
incomplete attempt at this). This would avoid issues like this one that are
caused by hacky workarounds. However, this approach has some downsides as well,
and the best approach is yet to be determined.

In the meantime, this PR fixes the bug by delaying def creation for anon consts
whose bodies are macro invocations until after we expand the macro and know
what is inside it. This is accomplished by adding information to create the
anon const's def to the data in `Resolver.invocation_parents`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10411081372/job/28834297876) [(plain)](https://github.com/rust-lang/rust/commit/d04b5dde5e023df9e34dac498343923751ad1457/checks/28834297876/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---

running 218 tests
........................................................................................  88/218
........................................................................................ 176/218
............2024-08-15T22:20:12.773348Z ERROR compiletest::runtest: fatal error, panic: "crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge."
[crashes] tests/crashes/128016.rs ... F
.............................

failures:
failures:

---- [crashes] tests/crashes/128016.rs stdout ----

error: crashtest no longer crashes/triggers ICE, horray! Please give it a meaningful name, add a doc-comment to the start of the test explaining why it exists and move it to tests/ui or wherever you see fit. Adding 'Fixes #<issueNr>' to your PR description ensures that the corresponding ticket is auto-closed upon merge.
thread '[crashes] tests/crashes/128016.rs' panicked at src/tools/compiletest/src/runtest.rs:368:18:
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
```

</details>

BoxyUwU reviewed (changes requested):
Can you add a test for the version of the ICE that refers to an in scope generic parameter:
```rust
macro_rules! len {
    ($x:ident) => {
        $x
    };
}

fn bar<const N: usize>() {
    let val: [bool; len!(N)] = [true; N];
}
```
And also the one with `Res::Err`
```rust
macro_rules! len {
    () => {
        target
    };
}

fn main() {
    let val: [bool; len!()] = [];
}
```

BoxyUwU commented:
cc @cjgillot

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10411889014/job/28836661798) [(plain)](https://github.com/rust-lang/rust/commit/9549c3aafa8bc7ba70678b91a331e02f6332a866/checks/28836661798/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.693 Building wheels for collected packages: reuse
#13 2.694   Building wheel for reuse (pyproject.toml): started
#13 2.943   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.945   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.945   Stored in directory: /tmp/pip-ephem-wheel-cache-sd7on957/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.947 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.352 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.352 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.903 Collecting virtualenv
#13 3.903 Collecting virtualenv
#13 3.940   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.010      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 84.8 MB/s eta 0:00:00
#13 4.062 Collecting platformdirs<5,>=3.9.1
#13 4.067   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.100 Collecting filelock<4,>=3.12.2
#13 4.122 Collecting distlib<1,>=0.3.7
#13 4.126   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.133      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 93.7 MB/s eta 0:00:00
#13 4.133      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 93.7 MB/s eta 0:00:00
#13 4.213 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.400 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      208832 kB
DirectMap2M:     9228288 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/2c93fabd98d2c183bcb3afed1f7d51b2517ac5ed/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-2c93fabd98d2c183bcb3afed1f7d51b2517ac5ed-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 30.26s
##[endgroup]
fmt check
Diff in /checkout/compiler/rustc_resolve/src/def_collector.rs:39:
     /// we need to wait until we know what the macro expands to before we create the def for
     /// the anon const. That's because we lower some anon consts into `hir::ConstArgKind::Path`,
     /// which don't have defs.
+    ///
+    ///
     /// See [`Self::visit_anon_const`].
     pending_anon_const_info: Option<PendingAnonConstInfo>,
     impl_trait_context: ImplTraitContext,
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_data_structures/src/binary_search_util/mod.rs" "/checkout/compiler/rustc_data_structures/src/binary_search_util/tests.rs" "/checkout/compiler/rustc_data_structures/src/svh.rs" "/checkout/compiler/rustc_data_structures/src/fingerprint.rs" "/checkout/compiler/rustc_data_structures/src/unord.rs" "/checkout/compiler/rustc_data_structures/src/small_c_str/tests.rs" "/checkout/compiler/rustc_data_structures/src/frozen.rs" "/checkout/compiler/rustc_resolve/src/macros.rs" "/checkout/compiler/rustc_resolve/src/imports.rs" "/checkout/compiler/rustc_resolve/src/build_reduced_graph.rs" "/checkout/compiler/rustc_resolve/src/errors.rs" "/checkout/compiler/rustc_resolve/src/check_unused.rs" "/checkout/compiler/rustc_resolve/src/late/diagnostics.rs" "/checkout/compiler/rustc_resolve/src/effective_visibilities.rs" "/checkout/compiler/rustc_resolve/src/lib.rs" "/checkout/compiler/rustc_resolve/src/def_collector.rs" "/checkout/compiler/rustc_resolve/src/diagnostics.rs" "/checkout/compiler/rustc_resolve/src/ident.rs" "/checkout/compiler/rustc_resolve/src/rustdoc.rs" "/checkout/compiler/rustc_resolve/src/late.rs" "/checkout/compiler/rustc_monomorphize/src/errors.rs" "/checkout/compiler/rustc_monomorphize/src/collector.rs" "/checkout/compiler/rustc_monomorphize/src/polymorphize.rs" "/checkout/compiler/rustc_monomorphize/src/collector/move_check.rs" "/checkout/compiler/rustc_monomorphize/src/lib.rs" "/checkout/compiler/rustc_monomorphize/src/partitioning.rs" "/checkout/compiler/rustc_monomorphize/src/util.rs" "/checkout/compiler/rustc_codegen_llvm/src/va_arg.rs" "/checkout/compiler/rustc_codegen_llvm/src/errors.rs" "/checkout/compiler/rustc_codegen_llvm/src/base.rs" "/checkout/compiler/rustc_codegen_llvm/src/asm.rs" "/checkout/compiler/rustc_codegen_llvm/src/intrinsic.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm_util.rs" "/checkout/compiler/rustc_codegen_llvm/src/type_of.rs" "/checkout/compiler/rustc_codegen_llvm/src/attributes.rs" "/checkout/compiler/rustc_codegen_llvm/src/common.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/diagnostic.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/mod.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/ffi.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/archive_ro.rs" "/checkout/compiler/rustc_codegen_llvm/src/context.rs" "/checkout/compiler/rustc_codegen_llvm/src/allocator.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/owned_target_machine.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/archive.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/profiling.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/write.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/lto.rs" "/checkout/compiler/rustc_codegen_llvm/src/type_.rs" "/checkout/compiler/rustc_codegen_llvm/src/declare.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/utils.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/mod.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata/type_map.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata/enums/mod.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata/enums/native.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata/enums/cpp_like.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs" "/checkout/compiler/rustc_codegen_llvm/src/lib.rs" "/checkout/compiler/rustc_codegen_llvm/src/consts.rs" "/checkout/compiler/rustc_codegen_llvm/src/coverageinfo/mapgen.rs" "/checkout/compiler/rustc_codegen_llvm/src/coverageinfo/mod.rs" "/checkout/compiler/rustc_data_structures/src/intern.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Aug 15 23:23:12 UTC 2024
  network time: Thu, 15 Aug 2024 23:23:12 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10412056270/job/28837108079) [(plain)](https://github.com/rust-lang/rust/commit/ef33827a7e85ff9aed3ca19e78917cad0fb87a30/checks/28837108079/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.879 Building wheels for collected packages: reuse
#16 2.880   Building wheel for reuse (pyproject.toml): started
#16 3.130   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.131   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.131   Stored in directory: /tmp/pip-ephem-wheel-cache-igwt3z48/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.133 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.529 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.530 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
    Checking rustc_lint v0.0.0 (/checkout/compiler/rustc_lint)
error: unresolved link to `Self::visit_anon_const`
  --> compiler/rustc_resolve/src/def_collector.rs:43:15
   |
43 |     /// See [`Self::visit_anon_const`].
   |               ^^^^^^^^^^^^^^^^^^^^^^ the struct `DefCollector` has no field or associated item named `visit_anon_const`
   = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`

error: could not document `rustc_resolve`
error: could not document `rustc_resolve`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Thu Aug 15 23:58:24 UTC 2024
  network time: Thu, 15 Aug 2024 23:58:24 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

petrochenkov commented:
Upd: I started reviewing this stuff today, starting from https://github.com/rust-lang/rust/pull/125915.
I'm generally skeptical about moving any def collection to later stages (except for leaf `DefId`s perhaps), but I need to read the relevant PRs first to say more.

petrochenkov commented:
I think I disagree with both this fix and the "long term fix" from https://github.com/rust-lang/rust/issues/129247.
I have left some comments in https://github.com/rust-lang/rust/issues/129023.

https://github.com/rust-lang/rust/pull/125915 didn't need to change anything in `DefId` creation and parenting.
The `AnonConst` def ids (or whole constants) can still be created for `ConstArgKind::Path` const expressions, it doesn't mean that they should be actually used in type checking and const evaluation for those expressions.

If `min_generic_const_args` goes further and supports arbitrary paths, then there will be situations in which it's not clear whether it is a "type system constant" even at AST lowering time.
E.g. in
```rust
Type1::Assoc<Type2::Assoc>
```
we do not know whether `Type2::Assoc` is an associated type or constant (or even enum variant constructor) because it's not resolved yet, and it cannot be determined from the expected `Type1::Assoc`'s generic parameters as well because `Type1::Assoc` is not yet resolved either.

I'd prefer https://github.com/rust-lang/rust/pull/125915 to be reverted and then re-landed with the HIR representation changes (`hir::{ConstArg, ConstArgKind}`), but without the `DefId` creation and parenting changes.

petrochenkov commented:
"Literature" list:
- https://hackmd.io/@BoxyUwU/BJ6_bfmD0 The plan for min_generic_const_args
- https://github.com/rust-lang/rust/pull/125915
- https://github.com/rust-lang/rust/pull/128016 
- https://github.com/rust-lang/rust/pull/128760
- https://github.com/rust-lang/rust/pull/128844
- https://github.com/rust-lang/rust/pull/128901
- https://github.com/rust-lang/rust/pull/129023
- https://github.com/rust-lang/rust/pull/129137
- https://github.com/rust-lang/rust/pull/129244
- https://github.com/rust-lang/rust/pull/129247
- https://github.com/rust-lang/rust/pull/129246

camelid commented:
> #125915 didn't need to change anything in `DefId` creation and parenting.
> The `AnonConst` def ids (or whole constants) can still be created for `ConstArgKind::Path` const expressions, it doesn't mean that they should be actually used in type checking and const evaluation for those expressions.

I was under the impression that we have a HIR validation check that ensures all `DefId`s are used. Is that not right? In any case, it would seem unfortunate to create defs that don't end up being used. It's also worth pointing out that we already did (before my PR) def creation in ast_lowering for some anon consts that were not parsed as such, like const paths parsed as type paths. Finally, IIRC we need to delay def creation to properly do query feeding for anon consts, though @BoxyUwU knows much more about this than I. I'm just not sure that it's as simple as moving the def creation earlier -- though I could be wrong.

> I'd prefer #125915 to be reverted and then re-landed with the HIR representation changes (`hir::{ConstArg, ConstArgKind}`), but without the `DefId` creation and parenting changes.

If we do decide to move def creation earlier, creating unused defs in the process, then I don't see the point of reverting the whole PR. We can just make a targeted fix to the def creation parts rather than redoing the whole thing.

On that note, would you be ok with landing this _current_ PR provided that we work to figure out a longer-term solution that is effective and satisfies all parties? It would be nice to get this bug fixed on nightly without the churn of reverting #125915 and going back to square one.

BoxyUwU commented:
I don't think we have such a check off the top of my head but I *do* think that creating needless DefIds is both hacky and will not work since we will have ICEs from calling queries on those DefId's which will have unexpected hir nodes associated with them (e.g. an expression instead of an anon const). This would happen when encoding queries for `AnonConst` defids.

I do not really understand the opposition to landing this PR, it gives us a fully correct DefId tree since all items defined in anon consts will have the anon const as the parent. It also does not result in having any extra `DefId`s defined which would be problematic ICE wise and also annoying when looking at debug output since we will wind up with `{constant#1}` when there is only a single constant.

I think it makes far more sense to land #129246 which will fix the immediate beta regression indefinitely, then allow this PR to be reviewed and merged which will solve things long term. As far as I can tell there is ~no downside to this PR? The only issues I've heard of with the current approach to creating defs for anon consts is that a "fully capable" def tree is desired and this PR allows for this.

I agree with @camelid that reverting the original PR seems like needless churn. IMO it is far too strong of a reaction considering the fairly minor issues with the current scheme.

BoxyUwU commented:
I don't entirely know if #129247 is what we want long term. I think if we can end up with an "obviously" correct DefId tree then that's better (which this PR does).

petrochenkov commented:
>I do not really understand the opposition to landing this PR

Because it layers more hacks on top of #125915's hacks and increases the amount of changes that I'd like to revert.
I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.

cjgillot commented:
Reverting #125915 agrees with the general policy to revert PRs that cause high impact bugs or perf regressions, to re-land them better later. 

On the general direction, we should strive for the simpler code. This PR introduces special cases in the def collector, to accommodate special cases in ast lowering. 

This makes me wonder: is ast lowering the right place to have the special case. Or is hir->ty lowering ? 

I'd be interested in reading the design you have in mind long term for gce feature, to understand where are your tradeoffs and what constraints you battle against and should be lifted.

BoxyUwU commented:
I agree that in general a policy of "revert then reland makes sense" I *tried* making a revert PR and it has merge conflicts which is why instead I filed #129246 which was easier and would also give good assurance that there shouldn't be any regressions (and also if there *is* a mistake in that PR or the revert, I think it will be easier to justify a backport of "add a feature gate check" than "touch this random compiler logic to make it not scuffed from a merge conflcit")

Moving hacks to hir ty lowering doesn't help anything IMO, I would *like* for anon consts to have their defs created there since it will help with the `type_of` feeding hack. That wouldn't change the fact that `DefCollector` shouldn't be creating the `DefId`s for things we either don't want to create or *do* want to create later on. It also is more complex since iirc `tcx.create_def` to create defids that should have bodies is a lot more error prone.

petrochenkov commented:
>I don't think we have such a check off the top of my head but I *do* think that creating needless DefIds is both hacky and will not work since we will have ICEs from calling queries on those DefId's which will have unexpected hir nodes associated with them (e.g. an expression instead of an anon const). This would happen when encoding queries for `AnonConst` defids.

Always creating `DefId`s (maybe even for all generic arguments, to avoid the "we already did def creation in ast_lowering for some anon consts" bit) would be the simplest alternative.
*Not* creating `DefId`s in some cases would be an optimization that increases complexity, and as with any optimizations it should first be proven that it is beneficial.
(At metadata encoding we already know precisely which def ids actually required anon constants and which not, that knowledge can be used to skip encoding.)

>Moving hacks to hir ty lowering doesn't help anything IMO, I would *like* for anon consts to have their defs created there since it will help with the `type_of` feeding hack.

I think that would be the second viable strategy, directly opposite to the early def id creation in def collector.
We already create artificial `static`s late in compilation similarly to this (see `intern_as_new_static`).
Such nodes do not have any nontrivial HIR (all necessary information about them is fed through queries), and don't participate in any nontrivial parenting relations, neither in `HirId` nor in `DefId` sense.
They can also be created exactly when necessary, when the constant expression is complex enough, including the `Type1::Assoc<Type2::Assoc>` cases.

So in both "create def ids very early" and "create def ids very late" we won't have approximations and hacks in def collector and AST lowering.
In the first case it's because we create them for everything, and in the second case because we have full information and can create them exactly when necessary.

camelid commented:
If we create defs very late, doesn't that run into the same issue of messing up parenting information for children of anon consts? I'm not sure I follow how it resolves the issue.

petrochenkov commented:
If the constant is a leaf side node that doesn't parent any HIR and doesn't parent any DefIds either, like `intern_as_new_static` static, then there's no mismatch.

BoxyUwU commented:
anon consts *do* parent other defs though? I don't understand the solution you're imagining. If it can be fine to create defs in hir ty lowering then I can't see how it wouldn't be possible to construct that same parent tree in ast lowering.

petrochenkov commented:
I mean it's we who decide what parents what, the scheme just needs to be self-consistent and preferably convenient to work with.
Right now anonymous constant parent HIR and inner def ids, but a different scheme similar to `intern_as_new_static` is likely also possible.

As I said I see two potentially viable, but opposite schemes that are relatively simple and not hacky - https://github.com/rust-lang/rust/pull/129137#issuecomment-2307477254, with their proc and cons.
Not being integrated into def id tree is generally not good, but for anon consts specifically that maybe be not too bad, because similarly to `intern_as_new_static` statics they are fully imaginary and don't exist in source code.

petrochenkov commented:
>I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.

I won't be able to do this today or tomorrow, unfortunately.
So I won't block this PR or https://github.com/rust-lang/rust/pull/129246 further, but I wrote enough details here and in https://github.com/rust-lang/rust/issues/129023 for you to think about the larger picture for `min_generic_const_args`, and maybe reconsider the direction.

bors commented:
:umbrella: The latest upstream changes (presumably #129521) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

camelid commented:
> > I'll try to implement my alternative on Saturday and see whether it can be done without introducing even worse hacks.
> 
> I won't be able to do this today or tomorrow, unfortunately. So I won't block this PR or #129246 further, but I wrote enough details here and in #129023 for you to think about the larger picture for `min_generic_const_args`, and maybe reconsider the direction.

Okay, sounds good. I definitely agree with you on the importance of ensuring we're comfortable with the implementation and that it won't introduce more hacky code than necessary.

bors commented:
:umbrella: The latest upstream changes (presumably #129841) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

camelid committed:
Re-enable `ConstArgKind::Path` lowering by default

...and remove the `const_arg_path` feature gate as a result. It was only
a stopgap measure to fix the regression that the new lowering introduced
(which should now be fixed by this PR).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10757497107/job/29831592260) [(plain)](https://github.com/rust-lang/rust/commit/7fa114c92bffd675b7702b3123482e7838780d70/checks/29831592260/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.854 Building wheels for collected packages: reuse
#16 2.855   Building wheel for reuse (pyproject.toml): started
#16 3.099   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.100   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.100   Stored in directory: /tmp/pip-ephem-wheel-cache-nv61_f0z/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.103 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.516 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.517 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
    Checking expect-test v1.5.0
error: unused variable: `qpath`
   --> compiler/rustc_middle/src/ty/consts.rs:299:13
    |
299 |             qpath @ hir::QPath::Resolved(
    |             ^^^^^ help: if this is intentional, prefix it with an underscore: `_qpath`
    = note: `-D unused-variables` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(unused_variables)]`

error: could not compile `rustc_middle` (lib) due to 1 previous error
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10766268896/job/29851756917) [(plain)](https://github.com/rust-lang/rust/commit/bbf0477c3e34389eaa95d74b963e2dccd95524bf/checks/29851756917/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:fb97a1d6377f6cf2227825318ca4bbde3889e0c420746f5a03ba46a07e9a862725c26a09b9fc49a0d129ebd75935d3f6cd19acf41cc4267a6846fd4aa574b12c:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

BoxyUwU commented:
You'll need to delete the `tests/ui/feature-gates/feature-gate-const-arg-path.rs` test

bors commented:
:umbrella: The latest upstream changes (presumably #130249) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

BoxyUwU commented:
@bors r+ rollup=never

Merging this to unblock more work on `min_generic_const_args` .

I don't consider "dont create extraneous defids" to be solely an optimisation as it will affect diagnostics output as anon const's will be misnumbered (e.g. we'll see `foo::{constant#1}` in output when there is only a single anon const). Having this increase in complexity in `DefCollector` seems logical to me since fundamentally what we are trying to do is not represent const args with anon consts in some cases that are only able to be determined *after* macro expansion.

Unconditionally creating DefIds and introducing hacks to defid encoding to skip these "fake" DefIds for anon consts might work (with the caveat that it'll make diagnostics and debugging worse(?)). However I think it places complexity in the compiler where it is not necessarily expected or relevant. 

After this lands we should be able to remove a lot of the DefId creation in ast lowering of const arguments by always representing things as `ConstArg::Path` instead of having the weird approximation in DefCollector of what needs a `DefId`. This would leave us in a final position where the only times we create `DefId`s for const args is:
- `DefCollector`
- `rustc_legacy_const_generics` handling which only knows that a function argument is actually a const argument after name resolution has been performed (unrelated but at some point we should fix the ICEs around having nested bodies in these arguments)

Hopefully once we have made it so that the DefId creation in DefCollector is not an under approximation of what defs we *actually* need the ast lowering logic will be less "hacky" and the state of all this will seem acceptable.

bors commented:
:pushpin: Commit e0bd01167e86d07c03e8ddd2bb0a25f689a2a7f5 has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU e0bd01167e86d07c03e8ddd2bb0a25f689a2a7f5 -->
<!-- homu: {"type":"Approved","sha":"e0bd01167e86d07c03e8ddd2bb0a25f689a2a7f5","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e0bd01167e86d07c03e8ddd2bb0a25f689a2a7f5 with merge d3a8524e80244d66968ab4522d31145ccba610f4...
<!-- homu: {"type":"BuildStarted","head_sha":"e0bd01167e86d07c03e8ddd2bb0a25f689a2a7f5","merge_sha":"d3a8524e80244d66968ab4522d31145ccba610f4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10841264537/job/30088681130)
Approved by: BoxyUwU
Pushing d3a8524e80244d66968ab4522d31145ccba610f4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"BoxyUwU","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10841264537/job/30088681130"},"merge_sha":"d3a8524e80244d66968ab4522d31145ccba610f4"} -->

rust-timer commented:
Finished benchmarking commit (d3a8524e80244d66968ab4522d31145ccba610f4): [comparison URL](https://perf.rust-lang.org/compare.html?start=adaff5368b0c7b328a0320a218751d65ab1bba97&end=d3a8524e80244d66968ab4522d31145ccba610f4&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=adaff5368b0c7b328a0320a218751d65ab1bba97&end=d3a8524e80244d66968ab4522d31145ccba610f4&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.4%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.9%, -0.5%] | 6     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 2     |
| All ❌✅ (primary)                 | -0.2% | [-0.9%, 0.4%]  | 12    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=adaff5368b0c7b328a0320a218751d65ab1bba97&end=d3a8524e80244d66968ab4522d31145ccba610f4&stat=max-rss)
<details>
<summary>Results (primary 0.5%, secondary 0.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [0.6%, 2.5%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 1.7%  | [1.3%, 3.0%]   | 4     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-1.1%, -0.8%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-2.0%, -1.1%] | 2     |
| All ❌✅ (primary)                 | 0.5%  | [-1.1%, 2.5%]  | 8     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=adaff5368b0c7b328a0320a218751d65ab1bba97&end=d3a8524e80244d66968ab4522d31145ccba610f4&stat=cycles:u)
<details>
<summary>Results (secondary 7.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 7.9% | [7.9%, 7.9%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=adaff5368b0c7b328a0320a218751d65ab1bba97&end=d3a8524e80244d66968ab4522d31145ccba610f4&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.1%, 0.7%]   | 36    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-1.0%, -0.0%] | 24    |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-1.0%, 0.7%]  | 60    |
</details>

**Bootstrap**: 756.349s -> 757.342s (0.13%)
**Artifact size**: 341.21 MiB -> 341.24 MiB (0.01%)
